### DISCUSSION TITLE
Restore a specific state?

### DISCUSSION THREAD

**Discussion by SchererM99** [PKU-Alignment/safety-gymnasium#121] [Q&A]
Hello! Is it possible to restore/load a specific state that was encountered during training? I have already found that in [world.py](https://github.com/PKU-Alignment/safety-gymnasium/blob/main/safety_gymnasium/world.py) functions to get and set a state are defined. I believe those, at least partly, can be used to reset the simulation. 

My question therefore is if a) these functions can be savely used to reset the whole simulation or if something will be missing. 
And b) if a similar functionality already exists to restore/load the whole environment (maybe even load e.g. a specific starting state that was not necessarily encountered before)

Thanks in advance!

**Comment by muchvo** [comment]
Hello, if your requirement is to save and restore specific environments, could you please check if the content in issue #95  addresses your concern? If you have any further questions, please feel free to ask.

**Reply by SchererM99** [reply]
Thank you for the quick answer! I already found that issue but my concern is that creating a deepcopy may take up a lot of memory during my experiments. Therefore I would prefer to just keep some information and "dynamically" restore an environment. But if that is not easily achievable the deepcopies should suffice. 

For my other question regarding the initial states. Is it possible to make the agent spawn in a specific x-y coordinate and orientation without creating a custom task? For my use case it would be extremly beneficial to create an experience starting from different coordinates and orientations by selecting those where more training is needed first.

**Reply by muchvo** [reply]
Certainly, you just need to save and then set each of the fields listed [here](https://mujoco.readthedocs.io/en/latest/APIreference/APItypes.html#mjtstate). Regarding modifying the agent's starting position and orientation, I believe you only need to manually set the corresponding attributes before each reset. Please refer to the discussion in discussion #112  for more details.

Feel free to raise any issues if you encounter any problems.
